docs/langref/relay_pattern.rst

```diff
@@ -0,0 +1,143 @@
+
+In the backend world, we could use the same machinery to build a higher level API using bring your own code generation. This API takes set of patterns describing your hardware capabilities and an external compiler, providing a relatively smooth heterogeneous experience out of the box.
+
+Recently there has been lots of discussion on similar issues in the community, and we wanted to gather feedback and hopefully collaborate on a design that can benefit everyone working in this space. This RFC focuses on the pattern language with future applications to come later.
+
+Examples
+========
+
+There are quite a few properties that are worth matching of operators below we examine how to match tree properties, and expand on some use cases that are not fully explored in the prototype. The first example is a simple case where we want to match one operator with a single input OR another operator with a single input, see the below diagram for a graphical representation and corresponding code::
+
+    def test_match_op_or():
+        is_add_or_sub = is_op('add') | is_op('subtract')
+        assert is_add_or_sub.match(relay.op.op.get("add"))
+        assert is_add_or_sub.match(relay.op.op.get("subtract"))
+
+The next example is a dense operation with any operator that is marked element-wise::
+
+    def test_no_match_attr():
+        op = is_op('nn.dense').has_attr("TOpPattern", K_ELEMWISE)
+        op_pat = op(wildcard(), wildcard())
+        x = relay.var('x')
+        y = relay.var('y')
+        assert not op_pat.match(relay.op.nn.dense(x, y))
+
+The next example is matching a diamond with two inputs at the top of the diamond::
+
+    def test_match_diamond():
+        # Pattern
+        is_conv2d = is_op('nn.conv2d')(is_input(), is_input())
+        path1 = is_op('nn.relu')(is_conv2d)
+        path2 = is_op('nn.leaky_relu')(is_conv2d)
+        diamond = is_op('add')(path1, path2)
+
+        # Expr
+        inp = relay.var('input')
+        weight = relay.var('weight')
+        conv2d = relay.op.nn.conv2d(inp, weight)
+        relu = relay.op.nn.relu(conv2d)
+        leaky_relu = relay.op.nn.leaky_relu(conv2d, alpha=0)
+        out = relu + leaky_relu
+
+        # Check
+        assert diamond.match(out)
+
+The final example we would like to match which is not yet implemented in the prototype is matching diamonds with a post-dominator relationship. Our plan is to embed dominator analysis as type of matching in the pattern language in order to allow for pattern matching with unknown topology.  This is important because we want to able to use the language to describe fuse patterns, like elementwise operations followed by a conv2d::
```

Comment:
```suggestion
The final example we would like to match which is not yet implemented in the prototype is matching diamonds with a post-dominator relationship. Our plan is to embed dominator analysis as type of matching in the pattern language in order to allow for pattern matching with unknown topology. This is important because we want to be able to use the language to describe fuse patterns, like elementwise operations followed by a conv2d::
```

Reply:
I think this sentence needs updating? Since it is no longer a prototype and dominator matching has been implemented @mbrookhart